flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/asyncstate/AsyncStateWordCount.java

```diff
@@ -0,0 +1,161 @@
+        "usage",
+        "This",
+        "Job",
+        "can",
+        "only",
+        "be",
+        "executed",
+        "with",
+        "async",
+        "state",
+        "enabled"
+    };
+
+    public static void main(String[] args) throws Exception {
+        MultipleParameterTool params = MultipleParameterTool.fromArgs(args);
+        int ttl = params.getInt("ttl", 0);
+        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+
+        // Use DataGenerator to generate data
+        final Random rnd = new Random();
+        GeneratorFunction<Long, String> generatorFunction =
+                index -> INPUTS[rnd.nextInt(INPUTS.length)];
+        DataGeneratorSource<String> source =
+                new DataGeneratorSource<>(
+                        generatorFunction,
+                        Long.MAX_VALUE,
+                        RateLimiterStrategy.perSecond(100),
+                        Types.STRING);
+
+        KeyedStream<String, String> streamSource =
+                env.fromSource(source, WatermarkStrategy.noWatermarks(), "Data Generator")
+                        .keyBy(e -> e);
+
+        FlatMapFunction<String, Tuple2<String, Long>> flatMapFunction = new WordCounter(ttl);
+
+        DataStream<Tuple2<String, Long>> mapper =
+                streamSource
+                        .enableAsyncState() // enable async state
+                        .flatMap(flatMapFunction)
+                        .setParallelism(3);
+
+        mapper.print().setParallelism(1);
+        env.execute("Async state Word Count example");
+    }
+
+    /** A simple word counter that uses state V2. */
+    public static class WordCounter extends RichFlatMapFunction<String, Tuple2<String, Long>> {
+
+        private transient ValueState<Integer> wordCounter;
+
+        private final long ttl;
+
+        public WordCounter(long ttl) {
+            this.ttl = ttl;
+        }
+
+        @Override
+        public void open(OpenContext context) {
+            ValueStateDescriptor<Integer> descriptor =
+                    new ValueStateDescriptor<>("Counter", BasicTypeInfo.INT_TYPE_INFO);
+            if (ttl > 0) {
+                StateTtlConfig ttlConfig =
+                        StateTtlConfig.newBuilder(Duration.ofMillis(ttl))
+                                .setUpdateType(StateTtlConfig.UpdateType.OnCreateAndWrite)
+                                .setStateVisibility(
+                                        StateTtlConfig.StateVisibility.ReturnExpiredIfNotCleanedUp)
+                                .build();
+                descriptor.enableTimeToLive(ttlConfig);
+            }
+            wordCounter = ((StreamingRuntimeContext) getRuntimeContext()).getValueState(descriptor);
+        }
+
+        @Override
+        public void flatMap(String value, Collector<Tuple2<String, Long>> out) throws Exception {
+            wordCounter
+                    .asyncValue()
+                    .thenAccept(
```

Comment:
I suggest using `thenApply` first to perform the null check and produce the value to update, then chained with another `thenAccept`

Reply:
Since `StateMachineExample` has `RichFlatMapFunction`, I removed this example and rewrite `StateMachineExample#StateMachineMapper` into state-v2 style.

Reply:
Make sense.